### PR TITLE
HHH-13900: added keepReference option to SQLDelete

### PR DESCRIPTION
**Pull request by richardbizik**
keepReference options allows hibernate to keep references for objects
that are being deleted. This behavior is wanted for soft delete
operations.

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/event/internal/DefaultDeleteEventListener.java [hunk 1]
```diff
@@ -281,7 +281,9 @@ protected final void deleteEntity(
 
 		new ForeignKeys.Nullifier(  entity, true, false, session, persister ).nullifyTransientReferences( entityEntry.getDeletedState() );
 		new Nullability( session ).checkNullability( entityEntry.getDeletedState(), persister, Nullability.NullabilityCheckType.DELETE );
-		persistenceContext.registerNullifiableEntityKey( key );
+		if( persister.getClassMetadata() != null && !persister.getClassMetadata().keepReference() ) {
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/event/internal/DefaultDeleteEventListener.java]: minor. A space should be inserted between `if` and `(`

File: hibernate-core/src/test/java/org/hibernate/test/annotations/delete/keepreference/DeathStar.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+public class DeathStar extends BaseEntity {
+
+    @NotNull
+	@OneToOne(optional = false, cascade = CascadeType.ALL, fetch = FetchType.EAGER)
+	private Vader vader;
+	@OneToMany(cascade = CascadeType.MERGE, orphanRemoval = true, mappedBy = "deathStar",fetch = FetchType.EAGER)
+	private Set<Trooper> troppers;
+    @NotNull
+    @OneToOne(optional = false, fetch = FetchType.EAGER)
+	private Universe universe;
```
Comment (NathanQingyangXu) [inline hibernate-core/src/test/java/org/hibernate/test/annotations/delete/keepreference/DeathStar.java]: there is vertical alignment issue above. Not a big deal though.

File: hibernate-core/src/test/java/org/hibernate/test/annotations/delete/keepreference/KeepReferenceTest.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import static org.junit.Assert.*;
+
+/**
+ * @author Richard Bizik
+ */
+public class KeepReferenceTest extends BaseCoreFunctionalTestCase {
+
+
+	@Test
+	@TestForIssue( jiraKey = "HHH-XXXXX")//TODO: assign jira key
```
Comment (NathanQingyangXu) [inline hibernate-core/src/test/java/org/hibernate/test/annotations/delete/keepreference/KeepReferenceTest.java]: I think we can eliminate this TODO for you have created the ticket

File: hibernate-core/src/test/java/org/hibernate/test/annotations/delete/keepreference/DeathStar.java (hunk 2 above)
Comment (richardbizik) [reply hibernate-core/src/test/java/org/hibernate/test/annotations/delete/keepreference/DeathStar.java]: I replaced all spaces by tabs in test package

**Review by sebersole — CHANGES_REQUESTED** [review]
I reverted your change, specifically in the delete listener, and re-ran your test - it passes.  Really need a test that shows this failing before your change and then passing with the change.

I also thought the test was confusing.  It may even have been styled after some of our own legacy tests, but those are confusing as well ;)

I also think we should just call this "soft delete" rather than "keep references" or "nullifiable".  At the very least, we should pick one or the other and use it consistently.  

I created 3659 on top of yours.  There are 2 commits...

The first reverts the delete listener change, so you can verify its not the test changes from my second commit.

The second commit (1) standardizes on calling this "soft delete" and (2) re-organizes the tests to be more clear (imo).

But we still need a test reproducing the the error without your change.

**Comment by richardbizik** [comment]
Hi, thanks for getting back to me. I tried your changes and you are right. Unfortunately its been quite some time and I cannot remember what was the exact setup when this happened to us and I could not reproduce it in tests. I only remember that this was the method that set the reference to null org.hibernate.engine.internal.ForeignKeys.Nullifier#nullifyTransientReferences
I hope I will have time to reproduce the issue in near future.

File: hibernate-core/src/main/java/org/hibernate/tuple/entity/EntityMetamodel.java [hunk 4]
```diff
@@ -139,6 +140,7 @@ public EntityMetamodel(
 		);
 
 		versioned = persistentClass.isVersioned();
+		keepReference = persistentClass.isCustomSQLDeleteNullifiable();
```
[the remark this answers is not in the document]
Comment (richardbizik) [reply hibernate-core/src/main/java/org/hibernate/tuple/entity/EntityMetamodel.java:143]: Nullifiable was just a naming taken from persistence context. `soft=true` sounds better

**Comment by gavinking** [comment]
Hibernate now has dedicated support for soft delete.

**Pull request closed without merging**